k-distribution/k-tutorial/1_basic/16_collections/README.md

```diff
@@ -1 +1,344 @@
+that contains such a map as a subset. The variable of sort `Map` will be bound
+to whatever bindings are left over (`.Map` if there are no bindings left over).
+
+Here is an example of a simple definition that implements a very basic
+variable declaration semantics using a `Map` to store the value of variables
+(`lesson-16-a.k`):
+
+```k
+module LESSON-16-A-SYNTAX
+  imports INT-SYNTAX
+  imports ID-SYNTAX
+
+  syntax Exp ::= Id | Int
+  syntax Decl ::= "int" Id "=" Exp ";" [strict(2)]
+  syntax Pgm ::= List{Decl,""}
+endmodule
+
+module LESSON-16-A
+  imports LESSON-16-A-SYNTAX
+  imports BOOL
+
+  configuration <T>
+                  <k> $PGM:Pgm </k>
+                  <state> .Map </state>
+                </T>
+
+  // declaration sequence
+  rule <k> D:Decl P:Pgm => D ~> P ...</k>
+  rule <k> .Pgm => . ...</k>
+
+  // variable declaration
+  rule <k> int X:Id = I:Int ; => . ...</k>
+       <state> STATE => STATE [ X <- I ] </state>
+
+  // variable lookup
+  rule <k> X:Id => I ...</k>
+       <state>... X |-> I ...</state>
+
+  syntax Bool ::= isKResult(K) [symbol, function]
+  rule isKResult(_:Int) => true
+  rule isKResult(_) => false [owise]
+endmodule
+```
+
+There are several new features in this definition. First, note we import
+the module `ID-SYNTAX`. This module is defined in `domains.md` and provides a
+basic syntax for identifiers. This syntax is only imported when parsing
+programs, not when parsing rules. Later in this lesson we will see how to
+reference specific concrete identifiers in a rule.
+
+Second, we introduce a single new function over the `Map` sort. This function,
+which is represented by the symbol
+`syntax Map ::= Map "[" KItem "<-" KItem "]"`, represents the map update
+operation. Other functions over the `Map` sort can be found in `domains.md`.
+
+Finally, we have used the `...` syntax on a cell containing a Map. In this
+case, the meaning of `<state>... Pattern ...</state>`,
+`<state>... Pattern </state>`, and `<state> Pattern ...</state>` are the same:
+it is equivalent to writing `<state> (Pattern) _:Map </state>`.
+
+Consider the following program (`a.decl`):
+
+```
+int x = 0;
+int y = 1;
+int a = x;
+```
+
+If we run this program with `krun`, we will get the following result:
+
+```
+<T>
+  <k>
+    .
+  </k>
+  <state>
+    a |-> 0
+    x |-> 0
+    y |-> 1
+  </state>
+</T>
+```
+
+Note that `krun` has automatically sorted the collection for you. This doesn't
+happen at runtime, so you still get the performance of a hash map, but it will
+help make the output more readable.
+
+### Exercise
+
+Create a sort `Stmt` that is a subsort of `Decl`. Create a production of sort
+`Stmt` for variable assignment. Feel free to use the syntax
+`syntax Stmt ::= Id "=" Exp ";"`. Write a rule that implements variable
+assignment using a map update function. Then write the same rule using a
+map pattern. Test your implementations with some programs to ensure they behave
+as expected.
```

Comment:
This one is a bit confusing to me. Should we be adding this in addition to the existing `Decl` assignment operator?

If it's meant that the `Decl` operator is variable declaration, and this is variable update, maybe we should refer to this as `variable update` or `variable assignment update` instead of just `variable assignment`?

I'm just confused by reading this if the intention is that it's added in addition to the existing assignment/declaration operator, or instead of.